src/concrete_solve.jl

```diff
@@ -334,6 +334,211 @@ function DiffEqBase._concrete_solve_adjoint(prob, alg,
+    # `_prob.u0`, this doesn't actually use more memory But it cleans up the
+    # implementation and makes `save_start` and `save_end` arg safe.
+    if typeof(sensealg) <: BacksolveAdjoint
+        # Saving behavior unchanged
+        ts = sol.t
+        only_end = length(ts) == 1 && ts[1] == _prob.tspan[2]
+        out = DiffEqBase.sensitivity_solution(sol, sol.u, ts)
+    elseif saveat isa Number
+        if _prob.tspan[2] > _prob.tspan[1]
+            ts = _prob.tspan[1]:convert(typeof(_prob.tspan[2]), abs(saveat)):_prob.tspan[2]
+        else
+            ts = _prob.tspan[2]:convert(typeof(_prob.tspan[2]), abs(saveat)):_prob.tspan[1]
+        end
+        # if _prob.tspan[2]-_prob.tspan[1] is not a multiple of saveat, one looses the last ts value
+        sol.t[end] !== ts[end] && (ts = fix_endpoints(sensealg, sol, ts))
+        if cb === nothing
+            _out = sol(ts)
+        else
+            _, duplicate_iterator_times = separate_nonunique(sol.t)
+            _out, ts = out_and_ts(ts, duplicate_iterator_times, sol)
+        end
+
+        out = if save_idxs === nothing
+            out = DiffEqBase.sensitivity_solution(sol, _out.u, ts)
+        else
+            out = DiffEqBase.sensitivity_solution(sol,
+                                                  [_out[i][save_idxs]
+                                                   for i in 1:length(_out)], ts)
+        end
+        only_end = length(ts) == 1 && ts[1] == _prob.tspan[2]
+    elseif isempty(saveat)
+        no_start = !save_start
+        no_end = !save_end
+        sol_idxs = 1:length(sol)
+        no_start && (sol_idxs = sol_idxs[2:end])
+        no_end && (sol_idxs = sol_idxs[1:(end - 1)])
+        only_end = length(sol_idxs) <= 1
+        _u = sol.u[sol_idxs]
+        u = save_idxs === nothing ? _u : [x[save_idxs] for x in _u]
+        ts = sol.t[sol_idxs]
+        out = DiffEqBase.sensitivity_solution(sol, u, ts)
+    else
+        _saveat = saveat isa Array ? sort(saveat) : saveat # for minibatching
+        if cb === nothing
+            _saveat = eltype(_saveat) <: typeof(prob.tspan[2]) ?
+                      convert.(typeof(_prob.tspan[2]), _saveat) : _saveat
+            ts = _saveat
+            _out = sol(ts)
+        else
+            _ts, duplicate_iterator_times = separate_nonunique(sol.t)
+            _out, ts = out_and_ts(_saveat, duplicate_iterator_times, sol)
+        end
+
+        out = if save_idxs === nothing
+            out = DiffEqBase.sensitivity_solution(sol, _out.u, ts)
+        else
+            out = DiffEqBase.sensitivity_solution(sol,
+                                                  [_out[i][save_idxs]
+                                                   for i in 1:length(_out)], ts)
+        end
+        only_end = length(ts) == 1 && ts[1] == _prob.tspan[2]
+    end
+
+    _save_idxs = save_idxs === nothing ? Colon() : save_idxs
+
+    function adjoint_sensitivity_backpass(Δ)
+        function df(_out, u, p, t, i)
```

Comment:
we could also write it as `df(u,p,t,i)`. I think that would make it clearer that it's not mutating.